realm/src/androidTest/java/io/realm/RealmAsyncQueryTests.java

```diff
@@ -0,0 +1,1943 @@
+                    threadAssertionError[0] = e;
+
+                } finally {
+                    if (signalCallbackFinished.getCount() > 0) {
+                        signalCallbackFinished.countDown();
+                    }
+                }
+            }
+        });
+
+        // wait until the callback of our async query proceed
+        signalCallbackFinished.await(7, TimeUnit.SECONDS);
+
+        if (null != threadAssertionError[0]) {
+            // throw any assertion errors happened in the background thread
+            throw threadAssertionError[0];
+        }
+        executorService.shutdownNow();
+    }
+
+    // similar UC as #testFindAllAsyncBatchUpdate using 'findAllSorted'
+    public void testFindAllSortedAsyncBatchUpdate() throws Throwable {
+        final CountDownLatch signalCallbackFinished = new CountDownLatch(2);
+        final Throwable[] threadAssertionError = new Throwable[1];
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        final AtomicInteger numberOfNotificationsQuery1 = new AtomicInteger(0);
+        final AtomicInteger numberOfNotificationsQuery2 = new AtomicInteger(0);
+        executorService.submit(new Runnable() {
+            @Override
+            public void run() {
+                Looper.prepare();
+                try {
+                    final Realm realm = openRealmInstance("testFindAllSortedAsyncBatchUpdate");
+                    final AtomicInteger numberOfIntercept = new AtomicInteger(1);
+                    final Handler handler = new HandlerProxy(realm.getHandler()) {
+                        @Override
+                        public boolean onInterceptMessage(int what) {
+                            switch (what) {
+                                case Realm.REALM_COMPLETED_ASYNC_QUERY: {
+                                    if (numberOfIntercept.getAndDecrement() > 0) {
+                                        // We advance the Realm so we can simulate a retry
+                                        // This is intercepted on the worker thread, we need to use
+                                        // the Realm on the caller thread
+                                        postAtFront(new Runnable() {
+                                            @Override
+                                            public void run() {
+                                                final CountDownLatch bgThreadLatch = new CountDownLatch(1);
+                                                new Thread() {
+                                                    @Override
+                                                    public void run() {
+                                                        Realm bgRealm = Realm.getInstance(realm.getConfiguration());
+                                                        bgRealm.beginTransaction();
+                                                        bgRealm.where(AllTypes.class).equalTo("co
```

Comment:
As a general thing, I don't think random numbers in tests - especially if seeded by time. It will make the test indeterministic and thereby harder to reproduce and debug.